### PR TITLE
refactor: send request with second-time confirmation

### PR DESCRIPTION
**Pull request by guanbinrui**
+ Use `chainId` to replace the `rpc` in overrides;
+ Need to confirm or reject each request by second time;

closes #

### CODE REVIEW

File: packages/maskbook/src/extension/background-script/EthereumServices/request.ts [hunk 1]
```diff
@@ -26,13 +58,38 @@ export async function request<T extends unknown>(requestArguments: RequestArgume
...
+    ) {
+        await WalletRPC.pushUnconfirmedRequest(payload_)
+        unconfirmedCallbackMap.set(payload_.id, callback)
+        if (popupsWindow) openPopupsWindow()
+        return
+    }
+    getSendMethod()(payload_, callback, overrides)
+}
+
+export async function confirmedRequest(payload: JsonRpcPayload) {
```
Comment (guanbinrui) [inline packages/maskbook/src/extension/background-script/EthereumServices/request.ts:82]:
```suggestion
export async function confirmRequest(payload: JsonRpcPayload) {
```

**Review by nuanyang233 — APPROVED** [review]

File: packages/maskbook/src/extension/background-script/EthereumServices/request.ts [hunk 2]
```diff
@@ -1,13 +1,41 @@
...
+    if (hasNativeAPI && nativeAPI) return INTERNAL_nativeSend
+    return INTERNAL_send
+}
+
+function getPayloadId(payload: JsonRpcPayload) {
+    return typeof payload.id === 'string' ? Number.parseInt(payload.id, 10) : payload.id
+}
+
+function isRpcNeedToBeConfirmed(method: EthereumMethodType) {
+    return [
```
Comment (septs) [inline packages/maskbook/src/extension/background-script/EthereumServices/request.ts:26]: extract the array to top-level as constant

File: packages/maskbook/src/extension/background-script/EthereumServices/request.ts [hunk 3]
```diff
@@ -1,13 +1,41 @@
...
+function getSendMethod() {
+    if (hasNativeAPI && nativeAPI) return INTERNAL_nativeSend
+    return INTERNAL_send
+}
+
+function getPayloadId(payload: JsonRpcPayload) {
+    return typeof payload.id === 'string' ? Number.parseInt(payload.id, 10) : payload.id
+}
+
+function isRpcNeedToBeConfirmed(method: EthereumMethodType) {
```
Comment (septs) [inline packages/maskbook/src/extension/background-script/EthereumServices/request.ts:25]: suggest, `rpc` all uppercase

```suggestion
function isRPCNeedToBeConfirmed(method: EthereumMethodType) {
```
```suggestion
function isRpcNeedToBeConfirmed(method: EthereumMethodType) {
```

File: packages/maskbook/src/extension/background-script/EthereumServices/request.ts [hunk 4]
```diff
@@ -26,13 +54,51 @@ export async function request<T extends unknown>(requestArguments: RequestArgume
...
     id += 1
-    await INTERNAL_send(
-        {
-            ...payload,
-            id,
-        },
-        callback,
-    )
+    const { providerType = currentProviderSettings.value } = overrides ?? {}
+    const payload_ = {
```
Comment (septs) [inline packages/maskbook/src/extension/background-script/EthereumServices/request.ts:61]: suggest, rename the variable name to `preparePayload`

```suggestion
    const preparePayload = {
```

File: packages/maskbook/src/extension/background-script/EthereumServices/send.ts [hunk 5]
```diff
@@ -54,22 +51,26 @@ function getChainIdFromPayload(payload: JsonRpcPayload) {
...
-        EthereumMethodType.PERSONAL_SIGN,
-        EthereumMethodType.ETH_SIGN_TYPED_DATA,
-        EthereumMethodType.ETH_DECRYPT,
-        EthereumMethodType.ETH_GET_ENCRYPTION_PUBLIC_KEY,
-        EthereumMethodType.ETH_SEND_TRANSACTION,
-    ].includes(payload.method as EthereumMethodType)
+function handleRecentTransaction(account: string, response: JsonRpcResponse | undefined) {
+    const hash = response?.result as string | undefined
+    if (typeof hash !== 'string') return
+    if (!/^0x([\dA-Fa-f]{64})$/.test(hash)) return
```
Comment (septs) [inline packages/maskbook/src/extension/background-script/EthereumServices/send.ts:57]: extract the regexp pattern to `@masknet/shared`?

CleanShot 2021-09-03 at 06 59 06@2x

Comment (septs) [reply packages/maskbook/src/extension/background-script/EthereumServices/send.ts:57]: or use `@masknet/web3-shared`?

File: packages/maskbook/src/extension/background-script/EthereumServices/send.ts [hunk 6]
```diff
@@ -255,36 +219,36 @@ export async function INTERNAL_send(
...
+        callback(null, response)
+        if (payload.method === EthereumMethodType.ETH_SEND_TRANSACTION) {
+            handleNonce(account, null, response)
+            handleRecentTransaction(account, response)
+        }
+    } catch (error) {
+        if (error instanceof Error) {
+            callback(error, undefined)
+            handleNonce(account, error, undefined)
+        }
```
Comment (septs) [inline packages/maskbook/src/extension/background-script/EthereumServices/send.ts:252]:
```suggestion
        } else {
            console.error('internal native send error', error)
        }
```

**Review by septs — APPROVED** [review]

**Review by nuanyang233 — APPROVED** [review]

**Pull request merged**